Nav Block - improving initial state by adding Block placeholder and ability to create from existing Pages.

getdave opened the pull request:
Closes https://github.com/WordPress/gutenberg/issues/18307.

This PR aims to improve the initial state of the Nav Block by adding a placeholder that affords the choice to:

1. Create a Menu from existing pages.
2. Create an empty Menu from scratch.

In future, this will be augmented to enable the creation of Menu from an existing Menu (but that is for another day!).

Note: if the Block _already_ has items within it (ie: you already created your menu) the placeholder will not show.

## Questions

* Would it be better to only request the Pages on demand rather than upfront? If so I'll need some advice on using `@wordpress/core-data` to achieve this as I can't seem to wrap my head around how I'd structure the code (cc @mtias ).

## How has this been tested?
Manual testing.

## Screenshots <!-- if applicable -->

![Screen Capture on 2019-11-08 at 14-34-38](https://user-images.githubusercontent.com/444434/68484386-152cc800-0235-11ea-94f3-ba630d4ecb6c.gif)



## Types of changes
New feature (non-breaking change which adds functionality).

## Checklist:
- [ ] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->

getdave committed:
Adds initial placeholder with basic state toggle

getdave committed:
Updates to only display placeholder if there are no existing Nav Items

getdave committed:
Update to tidy and simplify component organisation and logic

The component was becoming difficult to navigation. Organise into sections with clear comments, colocate similar items and simplify some conditionals.

getdave committed:
Updates visual design and copy

Addresses https://github.com/WordPress/gutenberg/issues/18307#issuecomment-551039990

getdave committed:
Updates variable naming in anticipation of other data types

mtias commented on packages/block-library/src/navigation-menu/edit.js:
```diff
@@ -108,14 +113,70 @@ function NavigationMenu( {
-	}, [ backgroundColor.class, textColor.class ] );
+	const handleCreateEmpty = () => {
+		setBlocksTemplate( null );
+		setInitialPlaceholder( false );
```
Why do we need all this state handling for the placeholder? It should just be a reflection of the state of other attributes:
- If block is empty (has no inner-blocks) => show placeholder.
- If block has only one inner block and it's empty => corresponds to start from scratch state.
- If block has multiple inner block elements => normal state.

getdave replied:
The placeholder showing can be a function of other props yes.

However, the buttons need to change the state of `innerBlocks`'s `template` prop.

 I needed someway to change the state of the component to tell `innerBlocks` what template it was going to use depending on which button was clicked:

1. No template.
2. Pages (which I'm going to load on demand shortly).

We can ditch the `[initialPlaceholder,setInitialPlaceholder]` but I felt we'd need the `blocksTemplate` especially when we'll offer different types of template in future (eg: Pages, Menu items...etc).

getdave replied:
States:

1. Placeholder
2. Created from Top Level Items
3. Created Empty

The `template` prop passed to `<innerBlocks>` determines the state of the UI. We change the `template` prop based on which option the user clicks. If there are existing innerBlocks we simply pass `null` as the `template` prop which allows the existing Blocks to appear.

mcsf replied:
> I felt we'd need the blocksTemplate especially when we'll offer different types of template in future (eg: Pages, Menu items...etc).

When the time comes to offer those, the model can be expanded. Until then, I expect this to hold:

>  It should just be a reflection of the state of other attributes

including in:

> I needed someway to change the state of the component to tell innerBlocks what template it was going to use depending on which button was clicked

getdave replied:
As I suspected (and failed to communicate effectively) this wasn't as simple as indicated above.

I discussed with @mtias and agreed that state was non-optimal in a Block context. I agree with this entirely and now better understand the reasons why.

However the issue of handling `InnerBlocks` was less simple. Our use case is unlike Media+Text or Columns because we do not utilise the `TemplateOptions` feature of `InnerBlocks` to render our placeholder. Whilst our version _looks_ visually similar, we have to render a `<Placeholder />` directly because `InnerBlocks` doesn't support a true placeholder attribute. As a result, we have a conditional render of either:

1. `<Placeholder />` - `InnerBlocks` is _not_ rendered here. We return early from render.
2. Block UI (ie: the Nav Block with all the controls...etc) - this is where `InnerBlocks` is rendered.

My point was that in the first case we do not render `InnerBlocks. As a result we need a way to change the "state" (not React `state`) of the component from `1` -> `2` when one of the `Button`s are clicked in the placeholder UI.

After discussing this at length, we decided that `dispatch`ing `replaceInnerBlocks` in the Button `onClick` handlers was the best way. This would trigger an update to `innerBlocks` causing `hasInnerBlocks` to return `true` therefore allowing the UI to transition from `1` -> `2`.

Apologies that I didn't communicate myself effectively. I've now updated the implementation and it's working without `useState` using the method described above.

getdave replied:
See updated PR desc screencapture.

mcsf commented on packages/block-library/src/navigation-menu/edit.js:
```diff
@@ -173,6 +240,7 @@ export default compose( [
 			orderby: 'id',
 		};
 		return {
+			hasExistingNavItems: !! hasExistingNavItems,
```
This can introduce confusion, since the "exported" prop is a Boolean while the inner one is an integer. Instead, the inner variable should be a Boolean from the start, i.e. in its assignment it should read `.length > 0` or equivalent.

getdave replied:
I've simplified this now.

mcsf commented on packages/block-library/src/navigation-menu/edit.js:
```diff
@@ -164,7 +225,13 @@ function NavigationMenu( {
-	withSelect( ( select ) => {
+	withSelect( ( select, ownProps, registry ) => {
+		const { clientId } = ownProps;
+		const { getBlocks } = registry.select( 'core/block-editor' );
```
Why consume `registry` instead of accessing `select` directly?

```js
withSelect( ( select, { clientId } ) => {
  const innerBlocks = select( 'core/block-editor' ).getBlocks( clientId );
  …
```

Additional notes:

1. Calling `select( store ).selectorName( … )` is more idiomatic in Gutenberg than extracting `const selectorName` beforehand.
1. The object destructuring near `{ clientId }` is just personal preference.

getdave replied:
@aduth Once told me that `getBlocks` was expensive and access directly from registry could help mitigate performance concerns. Indeed, this is what we do in the Columns Block (which is where I think he told me about this).

Happy to avoid if that's no longer the case or I misunderstood.

getdave replied:
Simplified and fixed other items. You're right it was additional "noise" and complexity for no tangible benefit to the readability of the code.

mcsf replied:
> aduth Once told me that `getBlocks` was expensive and access directly from registry could help mitigate performance concerns. Indeed, this is what we do in the Columns Block (which is where I think he told me about this).

Is this the comment in question? https://github.com/WordPress/gutenberg/pull/13899#discussion_r279876963

If so, and looking at the context, including the description of #11851, it sounds like this is about grabbing state to be used in actions via `withDispatch` and avoiding pulling that action-only state with another `withSelect`. Quote: "It is useful when you need to fetch some dynamic data from the store at the time when the event is fired, but at the same time, you never use it to render your component."

If I interpreted the above correctly, there is no benefit to consuming `registry`.

getdave replied:
Ah yes. You are right. I guess blindly follow advice isn't the best strategy even with the best of intentions. Thanks for flagging. Fixed.

mcsf commented on packages/block-library/src/navigation-menu/edit.js:
```diff
@@ -108,14 +113,70 @@ function NavigationMenu( {
+				<div className="wp-block-navigation-menu-placeholder__buttons">
+					<Button
+						className="wp-block-navigation-menu-placeholder__button"
+						isDefault={ true }
```
Minor tip: JSX allows `<Foo isDefault />` as shorthand for `<Foo isDefault={ true } />`.

getdave replied:
Noted.

mcsf commented on packages/block-library/src/navigation-menu/edit.js:
```diff
@@ -108,14 +113,70 @@ function NavigationMenu( {
+			[ attributes.backgroundColorCSSClass ]: attributes && attributes.backgroundColorCSSClass,
+			[ attributes.textColorCSSClass ]: attributes && attributes.textColorCSSClass,
+		}
+	);
```
These may be assigned after the placeholder return statement, for which they aren't needed.

getdave replied:
Good point. Fixed.

mcsf reviewed (changes requested):
I think the question of state in https://github.com/WordPress/gutenberg/pull/18363#discussion_r343694455 is important. Non-attribute state should be minimised, and so far I'm not convinced we need any.

The pitfalls of working with augmented state are illustrated by all the unexpected places where it breaks:

![gut-nav-menu](https://user-images.githubusercontent.com/150562/68421173-a4829e80-0195-11ea-84c7-4224dee23bc2.gif)

karmatosed commented:
One design feedback, could we reduce the spacing to be like other placeholders, please?

![image](https://user-images.githubusercontent.com/253067/68429473-0fd46c80-01a6-11ea-9985-7e0007b2e2d8.png)

talldan commented on packages/block-library/src/navigation-menu/edit.js:
```diff
@@ -108,14 +113,70 @@ function NavigationMenu( {
+				icon="menu"
+				label={ __( 'Navigation Menu' ) }
+			>
+				<p className="wp-block-navigation-menu-placeholder__instructions">
```
Placeholder has an `instructions` prop that'll handle rendering this.

getdave replied:
Silly of me to have missed this as I was taking note of the props on that component. Ironic that my classname of `___instructions` matches up the nomenclature the component author used for the prop!

Anyway, I've resolved this. Thanks!

talldan commented on packages/block-library/src/navigation-menu/edit.js:
```diff
@@ -151,7 +212,7 @@ function NavigationMenu( {
 				{ pages &&
 					<InnerBlocks
-						template={ defaultMenuItems ? defaultMenuItems : null }
+						template={ blocksTemplate ? blocksTemplate : null }
```
It looks like the default value of `blocksTemplate` is `null`, so probably no need for the ternary.

getdave replied:
No longer valid as we've ditched `template` all together 🎉

getdave committed:
Update don’t screen innerBlock unecessarily and simply bool prop.

Addresses

* https://github.com/WordPress/gutenberg/pull/18363#discussion_r343817558
* https://github.com/WordPress/gutenberg/pull/18363#discussion_r343820923

getdave commented on packages/block-library/src/navigation-menu/edit.js:
```diff
@@ -164,7 +225,13 @@ function NavigationMenu( {
+		const { getBlocks } = registry.select( 'core/block-editor' );
+
+		const innerBlocks = getBlocks( clientId );
+		const hasExistingNavItems = innerBlocks && innerBlocks.filter( ( block ) => block.name === 'core/navigation-menu-item' ).length;
```
I thought perhaps the Block Appender would end up in there. I was overthinking it. Simplified to just check on presence of inner blocks (or not!).

getdave committed:
Updates to simplify withSelect and idioms around this

See https://github.com/WordPress/gutenberg/pull/18363#discussion_r343821227

getdave committed:
Simplifies JSX booleans

getdave committed:
Move rendered Block render attributes to be assigned only when used

Addresses https://github.com/WordPress/gutenberg/pull/18363#discussion_r343827595

getdave commented:
> One design feedback, could we reduce the spacing to be like other placeholders, please?

@karmatosed Noted. [Thanks to Dan](https://github.com/WordPress/gutenberg/pull/18363#discussion_r344013173) we are now making use of more of the defaults provided by `<Placeholder />`. I've then adjusted some other spacing. 

Please note that the buttons are spaced by `14px` between them. This is:

* `4px`: space between the buttons on the Gallery Block placeholder.
* `10px` hoz padding value of a default button

I add the `10px` because the 2nd button is in `isLink` style which means it doesn't have any inner padding so I compensate by adding more spacing otherwise it looks squashed visually. 

Here's what it looks like up against Gallery:

![Screen Shot 2019-11-08 at 11 24 27](https://user-images.githubusercontent.com/444434/68473406-ceca6f80-021a-11ea-9655-030c9fe27cb5.png)


If you'd like to adjust further please do let me know.

getdave committed:
Update design to match Gallery Block

getdave commented:
I'm going to work on the state issues now.

mtias commented on packages/block-library/src/navigation-menu/editor.scss:
```diff
@@ -154,3 +154,17 @@ $color-control-label-height: 20px;
 	}
 }
+
+// Initial Placeholder
```
I feel this is probably something we should add in the placeholder component or not touch.

getdave replied:
Let me look into that.

getdave replied:
Updated this so that any buttons used in Placeholder now have some default margins.

getdave committed:
Updates implementation to remove state and utilise dispatch to update innerBlocks

After a long discussion with @mtias we realised that unlike other similar implementations (Media+Text and Columns) of a placeholder, this one couldn’t use the “Template Options” feature of <InnerBlocks> to achieve the custom design required. Instead it rendered a `<Placeholder />` directly based on conditions.

As a result, we now use the `replaceInnerBlocks` from `block-editor` to dispatch an update to innerBlocks manually when the placeholder buttons are clicked.

Addresses

* https://github.com/WordPress/gutenberg/pull/18363#discussion_r343694455
* https://github.com/WordPress/gutenberg/pull/18363#discussion_r343833490
* https://github.com/WordPress/gutenberg/pull/18363#discussion_r344013908

getdave committed:
Fix to avoid improper use of `registry` when `select` is most appropriate.

Addresses https://github.com/WordPress/gutenberg/pull/18363#discussion_r344144855

getdave commented:
> The pitfalls of working with augmented state are illustrated by all the unexpected places where it breaks:

@mcsf This now looks fixed as a result of my updated implementation.

![Screen Capture on 2019-11-08 at 14-38-41](https://user-images.githubusercontent.com/444434/68484641-91bfa680-0235-11ea-9b31-e6da3220212e.gif)

mcsf commented on packages/block-library/src/navigation-menu/edit.js:
```diff
@@ -64,6 +64,8 @@ function NavigationMenu( {
 		} );
 	}, [ backgroundColor.class, textColor.class ] );
 
+	const emptyNavItemBlock = createBlock( 'core/navigation-menu-item' );
```
I think it's best to invoke `createBlock( 'core/navigation-menu-item' )` inside `handleCreateEmpty`, otherwise we have an object as referenced by `emptyNavItemBlock` which can — later on, as this file evolves — mistakenly end up used in more than one place, leading to undetermined effects.

getdave replied:
Fixed.

mcsf reviewed (commented):
Thanks for these changes, it's looking much cleaner.

getdave committed:
Updates to move button styling directly to component.

Acting on advice from @mtias.

getdave committed:
Refactor to allow “Automatically Add” Inspector Control to remove Placeholder

getdave commented:
@mtias I've updated so that clicking `Automatically add new pages` in inspector controls when the Block is in "Placeholder" mode (only!) automatically adds all top-level pages to the Nav.

What should happen in the case that I have existing items in the Nav and then I check this setting? Is the expectation that all top level items get auto-added to the Block? What happens if I then un-check the setting - does it then remove those items?

karmatosed commented:
> If you'd like to adjust further please do let me know.

No that looks fine to me now, thanks.

karmatosed reviewed (approved):
Based on the design and to get this v1 out, approving.

mtias commented:
@getdave under certain conditions (load from scratch, add block, choose starting blank) you can get the "loading" message when choosing the empty option:

<img width="950" alt="Screenshot 2019-11-10 at 17 54 18" src="https://user-images.githubusercontent.com/548849/68547552-2adff000-03e3-11ea-81ea-bf4bce9f2e32.png">

getdave committed:
Fix bug where spinner was displayed even for empty menus

Addresses https://github.com/WordPress/gutenberg/pull/18363#issuecomment-552212619

getdave commented:
> @getdave under certain conditions (load from scratch, add block, choose starting blank) you can get the "loading" message when choosing the empty option:
> 

I've replicated this by 

1. Loading new blank Post
2. Throttling network speed in devtools to `Slow 3G`
3. Adding Nav Block and _immediately_ selecting use empty menu.

I've now fixed this issue.

getdave commented:
New bug:

1. Load a new blank Post
2. Throttle network speed in devtools to Slow 3G.
3. Add Nav Block and immediately select `Create from all top pages`.

If the Pages are yet to load then there is an error.

getdave committed:
Fix to only allow creation from Pages if available.

getdave committed:
Update to improve safeguards around creation of Menus from Pages before Page data is available

Previously we didn’t know when the pages query had resolved. Moreover, we were allowing the Inspector controls to create Menus from Pages before the data was available.

Add conditionals to cover these scenarios and DRY up selectors within `withSelect`

getdave commented:
Thanks to everyone who helped review this one. Appreciate your patience.

mtias commented:
Thanks for all the improvements!
